### ISSUE TITLE
Link rewriting provides false sense of security

### ISSUE THREAD

**Issue by ClearlyClaire** [mastodon/mastodon#11464]
The mechanism introduced in #11426 provides a false sense of security.

Indeed, it by only rewriting links which text visually resembles links, it will allow any fake link including zero-width/non-printable characters. We could try pre-processing the text by removing such characters, but it seems extremely difficult to come up with an exhaustive list of such characters.

Relying on a set of acceptable characters won't work either, because we want to allow custom link text made of words in any language, as to not mangle remote messages in a way that would make them lose meaning.

EDIT: The two options I see are, either go back to my original proposal, or show an alert when clicking on any link that does not match its target (even if they do not “look like” links)

**Comment by duxovni** [comment]
I'm also mildly concerned about link rewriting after thinking about it a bit longer. I'm imagining a troll writing a toot like

```
<a href="wholesome.website">nazis.com</a> is an awesome website!
```

A mastodon user sees

```
wholesome.website is an awesome website!
```

and boosts it, and now people on fediverse servers that don't implement link rewriting see the mastodon user boosting nazi content.

I like the idea of showing an alert when clicking on any link that doesn't match its target, but also allowing the user to build up a whitelist of trusted sites.

**Comment by duxovni** [comment]
Something like your original proposal could also be good, if we can figure out a way of displaying the real link destination that's accessible and not too cluttering.

**Comment by ThibG** [comment]
On my glitch-soc PR (which really is just another iteration of my original proposal), this looks like the following:

![image](https://user-images.githubusercontent.com/384364/62349542-59a31c00-b500-11e9-8721-9816f4c74f6c.png)

I do not think this is too cluttered, but maybe we could improve it further.

**Comment by clarfon** [comment]
Perhaps some form of UTS 39 might help? http://www.unicode.org/reports/tr39/

**Comment by ThibG** [comment]
TL;DR: implementing “detecting things that look like links* is vastly more complicated than the techniques described in UTS39, which are an unwieldy beasts with no readily-available JS  implementation

Not really. Most of UTS 39 assumes you can restrict valid input to an Identifier Profile, for instance, but we want to allow free-form text in any language as link description, which means we cannot do that. Of all those texts, we would then need to find out which *look like* links, which involves considerations similar to that of UTS39 (for instance, consumables glyphs), but since we are not comparing two strings nor can we restrict the used character set, this is vastly more complicated.

On a related note, UTS39 can be useful against homograph attacks on International Domain Names, which neither the current Mastodon code nor the related PR does. Indeed, doing that in JavaScript would be a *major* undertaking, especially doing that in an even remotely efficient way.
